File: programmers/level3/섬연결하기.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstring>

using namespace std;

int parent[101];

int Find(int x){
    if (x==parent[x]) return x;
    else {
        int y = Find(parent[x]);
        parent[x] = y;
        return y;
    }
}

void Union(int x,int y){
    y = Find(y);
    x = Find(x);
    if (x!=y){
        parent[y] = x;
    }
}

class Edge {
    public:
        int node[2];
        int distance;

        Edge(int a,int b,int distance){
            this->node[0] = a;
            this->node[1] = b;
            this->distance = distance;
        }

        bool operator<(const Edge &edge) const {
            return (this->distance) < (edge.distance);
        }
};

int solution(int n, vector<vector<int>> costs){
    
    vector <Edge> v;
    for (int i=0;i<costs.size();i++){
        v.push_back(Edge(costs[i][0], costs[i][1], costs[i][2]));
    }

    sort(v.begin(), v.end());
    for (int i=0;i<n;i++){
        parent[i] = i;
    }

    int answer = 0;
    for (int i=0;i<v.size();i++){
        if (Find(v[i].node[0]) != Find(v[i].node[1])){
;
            answer += v[i].distance;
            Union(v[i].node[0], v[i].node[1]);
        }
    }

    return answer;
}

int main(){
    cout << solution(4, {{0,1,1},{0,2,2},{1,2,5},{1,3,1},{2,3,8}}) << "\n";
    return 0;
}

File: programmers/level3/섬연결하기.py
parent = []


def find(u):
    if u != parent[u]:
        parent[u] = find(parent[u])
    return parent[u]


def union(u, v):
    root1 = find(u)
    root2 = find(v)

    parent[root2] = root1


def solution(n, costs):

    for i in range(n):
        parent.append(i)

    costs = sorted(costs, key=lambda x: x[2])

    edge_cnt = 0
    answer = 0

    while True:
        if edge_cnt == n-1:
            break

        u, v, cost = costs.pop(0)
        print(u, v, cost)
        print(edge_cnt)

        if find(u) != find(v):
            union(u, v)
            answer += cost
            edge_cnt += 1

    return answer


solution(4, [[0, 1, 1], [0, 2, 2], [1, 2, 5], [1, 3, 1], [2, 3, 8]])
